oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/CachedGroupMembershipReader.java

```diff
@@ -0,0 +1,209 @@
+            readGroupsFromCache(authorizable, groupPrincipals, this::loadGroupPrincipals);
+        } else {
+            loadGroupPrincipals(authorizable, groupPrincipals);
+        }
+    }
+
+    private void readGroupsFromCache(@NotNull Tree authorizableNode,
+                                     @NotNull Set<Principal> groups,
+                                     @NotNull BiConsumer<Tree, Set<Principal>> loader) {
+        long expirationTime = 0;
+        String authorizablePath = authorizableNode.getPath();
+        Tree principalCache = authorizableNode.getChild(CacheConstants.REP_CACHE);
+        if (principalCache.exists()) {
+            expirationTime = TreeUtil.getLong(principalCache, CacheConstants.REP_EXPIRATION, EXPIRATION_NO_CACHE);
+        }
+        long now = System.currentTimeMillis();
+        if (expirationTime > EXPIRATION_NO_CACHE && now < expirationTime) {
+            serveGroupsFromCache(authorizablePath, principalCache, groups);
+            return;
+        }
+
+        // need to load or serve stale if allowed
+        boolean mayServeStale = now - expirationTime < maxStale;
+        boolean updateCache;
+        Map<String, Long> updates = getCacheUpdateMap();
+        synchronized (updates) {
+            // anyone else updating that entry already?
+            Long exp = updates.get(authorizablePath);
+            updateCache = exp == null || expirationTime > exp;
+            if (updateCache) {
+                // we do it
+                updates.put(authorizablePath, expirationTime);
+            }
+        }
+
+        if (!updateCache && mayServeStale) {
+            LOG.debug("Another thread is updating the cache and we may serve stale");
+            serveGroupsFromCache(authorizablePath, principalCache, groups);
+            return;
+        }
+
+        if (updateCache) {
+            boolean cacheSuccess = false;
+            try {
+                loader.accept(authorizableNode, groups);
+                cacheGroups(authorizableNode, groups);
+                cacheSuccess = true;
+            } catch (AccessDeniedException | CommitFailedException e) {
+                LOG.debug("Failed to cache group membership: {}", e.getMessage());
+            } finally {
+                if (!cacheSuccess || expirationTime == 0) {
+                    synchronized (updates) {
+                        Long exp = updates.get(authorizablePath);
+                        if (Objects.equals(exp, expirationTime)) {
+                            updates.remove(authorizablePath);
+                        }
+                    }
+                }
+            }
+        } else {
+            // load but do not cache. happens when another thread is updating
+            // the cache and this thread is not allowed to serve stale
+            LOG.debug("Load but do not cache. Another thread is updating the cache and this thread is not allowed to serve stale");
```

Comment:
same applies here. if i were an outsider i would not understand what is 'stale' and what the implications are.